src/store/conversationsStore.js

```diff
@@ -122,6 +123,71 @@ const mutations = {
 	addConversation(state, conversation) {
 		Vue.set(state.conversations, conversation.token, conversation)
 	},
+
+	/**
+	 * Add new conversations
+	 *
+	 * @param {object} state the state
+	 * @param {object[]} conversations new conversations list
+	 */
+	addConversations(state, conversations) {
+		/**
+		 * Apply mutations to object based on new object:
+		 * - Add new properties
+		 * - Delete removed properties
+		 * - If withPropertiesUpdates - also update properties with primitive values (ignore Objects)
+		 *
+		 * @param {object} target target object
+		 * @param {object} newObject new object to get changes from
+		 * @param {boolean} [withPropertiesUpdates=false] also update properties with primitive values
+		 */
+		const applyObjectChanges = (target, newObject, withPropertiesUpdates = false) => {
+			// Delete removed properties
+			for (const key of Object.keys(target)) {
+				if (!Object.hasOwn(newObject, key)) {
+					del(target, key)
+				}
+			}
+			// Add new properties
+			for (const key of Object.keys(newObject)) {
+				if (!Object.hasOwn(target, key)) {
+					set(target, key, newObject[key])
+				}
+			}
+			// Update changed properties
+			if (withPropertiesUpdates) {
+				for (const [key, value] of Object.entries(target)) {
+					const newValue = newObject[key]
+					if (typeof value !== 'object' || typeof newValue !== 'object') {
+						// Update primitive property
+						set(target, key, newValue)
+					}
+				}
+			}
+		}
+
+		// Create a new conversation-by-Token object
+		const newConversations = {}
+		for (const conversation of conversations) {
+			newConversations[conversation.token] = conversation
+		}
+
+		// Add new conversations, remove old conversations
+		applyObjectChanges(state.conversations, newConversations, false)
+		// Update conversations
+		for (const [token, oldConversation] of Object.entries(state.conversations)) {
+			const newConversation = newConversations[token]
+			// If a conversation has activity - update conversation properties as well
+			if (oldConversation.lastActivity !== newConversation.lastActivity) {
+				applyObjectChanges(oldConversation, newConversation, true)
+				// If the conversation's last message was changed - update the last message
+				if (oldConversation.lastMessage && newConversation.lastMessage && oldConversation.lastMessage.id !== newConversation.lastMessage.id) {
+					applyObjectChanges(oldConversation.lastMessage, newConversation.lastMessage, true)
+				}
```

Comment:
It's the only `Object type` property for now, but to be more secure, could we make a recursive update here?
Or inside `applyObjectChanges()`?

Reply:
> but to be more secure, could we make a recursive update here?

Speaking about security, after your message I realized that there is a bug. I will fix it soon (now). Objects and arrays inside `lastMessage` were not updated.

---

> but to be more secure, could we make a recursive update here?

Well, on one hand, yes, on another no.

With the recursive update, it's a bit more complex to debug/control what is updating and how except setting the depth of blockLists. To see that "a conversation is updated" or "a lastMessage is updated".

But yep, with just a full recursive update we can make it universal and safe.

Both implementations are ready. What would you prefer?

Reply:
> we can make it universal and safe.

I'd vote for this one.

Just a thought, `lastMessage` property 'message' is one and only what we use in the component's render. Apart from that prop others are not used, and doubtly will change (only a reactions to last message, but that's not the reason to re-render a component)
So we could maybe rewrite a `lastMessage` object (and other 2-nd level object like it) completely to make it simple?